File: 0501-1000/0502_findMaximizedCapital.py
class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int: 
        n = len(profits)
        projects = sorted(list(zip(capital, profits)))
        heap = []
        unavailable = 0
        for i in range(k):
            while unavailable < n and projects[unavailable][0] <= w:
                heappush(heap, -projects[unavailable][1])
                unavailable += 1
            if not heap:
                break
            w -= heappop(heap)
        return w

# Alternative solution

class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        usableHeap = []
        reserveHeap = []

        for i in range(len(profits)):
            if capital[i] <= w:
                heappush(usableHeap, (-profits[i], capital[i]))
            else:
                heappush(reserveHeap, (capital[i], profits[i]))
    
        while k:
            if not usableHeap:
                break
                
            profit, capital = heappop(usableHeap)
            w -= profit
            k -= 1

            while reserveHeap and reserveHeap[0][0] <= w:
                capital, profit = heappop(reserveHeap)
                heappush(usableHeap, (-profit, capital))

        return w

# Alternative solution

class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        ipo = sorted([(c, p) for c, p in zip(capital, profits)], key = lambda x: [x[0], -x[1]])
        heap = []
        i = 0

        while k:
            if i < len(ipo) and ipo[i][0] <= w:
                heappush(heap, -ipo[i][1])
                i += 1
                continue
            
            if not heap:
                break

            w -= heappop(heap)
            k -= 1
                        
        return w

File: 0501-1000/0502_findMaximizedCapital.java
class Solution {
    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        int n = profits.length;
        int[][] combined = new int[n][2];

        for (int i = 0; i < n; i++) {
            combined[i][0] = profits[i];
            combined[i][1] = capital[i];
        }

        Arrays.sort(combined, (a, b) -> a[1] - b[1]);
        PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        int i = 0;

        while (i < profits.length && k > 0) {
            while (i < n && combined[i][1] <= w) {
                heap.offer(new int[] {-combined[i][0], combined[i][1]});
                i++;
            }

            if (heap.isEmpty()) {
                break;
            }

            if (!heap.isEmpty() && k > 0) {
                int[] pop = heap.poll();
                w -= pop[0];
                k--;
            }
        }

        while (!heap.isEmpty() && k > 0) {
            int[] pop = heap.poll();
            w -= pop[0];
            k--;
        }

        return w;
    }
}
